node/network/statement-distribution/src/lib.rs

```diff
@@ -0,0 +1,1376 @@
+/// via other means.
+#[derive(Default)]
+struct VcPerPeerTracker {
+	local_observed: arrayvec::ArrayVec<[Hash; VC_THRESHOLD]>,
+	remote_observed: arrayvec::ArrayVec<[Hash; VC_THRESHOLD]>,
+}
+
+impl VcPerPeerTracker {
+	// Note that the remote should now be aware that a validator has seconded a given candidate (by hash)
+	// based on a message that we have sent it from our local pool.
+	fn note_local(&mut self, h: Hash) {
+		if !note_hash(&mut self.local_observed, h) {
+			log::warn!("Statement distribution is erroneously attempting to distribute more \
+				than {} candidate(s) per validator index. Ignoring", VC_THRESHOLD);
+		}
+	}
+
+	// Note that the remote should now be aware that a validator has seconded a given candidate (by hash)
+	// based on a message that it has sent us.
+	//
+	// Returns `true` if the peer was allowed to send us such a message, `false` otherwise.
+	fn note_remote(&mut self, h: Hash) -> bool {
+		note_hash(&mut self.remote_observed, h)
+	}
+}
+
+fn note_hash(
+	observed: &mut arrayvec::ArrayVec<[Hash; VC_THRESHOLD]>,
+	h: Hash,
+) -> bool {
+	if observed.contains(&h) { return true; }
+
+	if observed.is_full() {
+		false
+	} else {
+		observed.try_push(h).expect("length of storage guarded above; \
+			only panics if length exceeds capacity; qed");
+
+		true
+	}
+}
+
+// knowledge that a peer has about goings-on in a relay parent.
+#[derive(Default)]
+struct PeerRelayParentKnowledge {
+	// candidates that the peer is aware of. This indicates that we can
+	// send other statements pertaining to that candidate.
+	known_candidates: HashSet<Hash>,
+	// fingerprints of all statements a peer should be aware of: those that
+	// were sent to the peer by us.
+	sent_statements: HashSet<(CompactStatement, ValidatorIndex)>,
+	// fingerprints of all statements a peer should be aware of: those that
+	// were sent to us by the peer.
+	received_statements: HashSet<(CompactStatement, ValidatorIndex)>,
+	// How many candidates this peer is aware of for each given validator index.
+	seconded_counts: HashMap<ValidatorIndex, VcPerPeerTracker>,
+	// How many statements we've received for each candidate that we're aware of.
+	received_message_count: HashMap<Hash, usize>,
+}
+
+impl PeerRelayParentKnowledge {
+	/// Attempt to update our view of the peer's knowledge with this statement's fingerprint based
+	/// on something that we would like to send to the peer.
+	///
+	/// This returns `None` if the peer cannot accept this statement, without altering internal
+	/// state.
+	///
+	/// If the peer can accept the statement, this returns `Some` and updates the internal state.
+	/// Once the knowledge has incorporated a statement, it cannot be incorporated again.
+	///
+	/// This returns `Some(true)` if this is the first time the peer has become aware of a
+	/// candidate with the given hash.
+	fn send(&mut self, fingerprint: &(CompactStatement, ValidatorIndex)) -> Option<bool> {
```

Comment:
`Option<bool>` contains no semantic information, and requires combinatorial mapping in documentation to make sense of it.

Reply:
Not sure I agree that it contains _no_ semantic information. `Option<bool>` is a poor man's trinary bit. I'd readily agree that a 3-way enum is often more useful.

Reply:
Yeah, 3-way enum is a little more idiomatic but also in this case feels like clutter.